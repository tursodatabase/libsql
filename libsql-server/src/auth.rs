use anyhow::{bail, Context as _, Result};
use axum::http::HeaderValue;
use libsql_replication::rpc::replication::NAMESPACE_METADATA_KEY;
use tonic::Status;

use crate::namespace::NamespaceName;

static GRPC_AUTH_HEADER: &str = "x-authorization";
static GRPC_PROXY_AUTH_HEADER: &str = "x-proxy-authorization";

/// Authentication that is required to access the server.
#[derive(Default)]
pub struct Auth {
    /// When true, no authentication is required.
    pub disabled: bool,
    /// If `Some`, we accept HTTP basic auth if it matches this value.
    pub http_basic: Option<String>,
    /// If `Some`, we accept all JWTs signed by this key.
    pub jwt_key: Option<jsonwebtoken::DecodingKey>,
}

#[derive(thiserror::Error, Debug)]
pub enum AuthError {
    #[error("The `Authorization` HTTP header is required but was not specified")]
    HttpAuthHeaderMissing,
    #[error("The `Authorization` HTTP header has invalid value")]
    HttpAuthHeaderInvalid,
    #[error("The authentication scheme in the `Authorization` HTTP header is not supported")]
    HttpAuthHeaderUnsupportedScheme,
    #[error("The `Basic` HTTP authentication scheme is not allowed")]
    BasicNotAllowed,
    #[error("The `Basic` HTTP authentication credentials were rejected")]
    BasicRejected,
    #[error("Authentication is required but no JWT was specified")]
    JwtMissing,
    #[error("Authentication using a JWT is not allowed")]
    JwtNotAllowed,
    #[error("The JWT is invalid")]
    JwtInvalid,
    #[error("The JWT has expired")]
    JwtExpired,
    #[error("The JWT is immature (not valid yet)")]
    JwtImmature,
    #[error("Authentication failed")]
    Other,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Authorized {
    pub namespace: Option<NamespaceName>,
    pub permission: Permission,
}

#[non_exhaustive]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Permission {
    FullAccess,
    ReadOnly,
}

/// A witness that the user has been authenticated.
#[non_exhaustive]
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Authenticated {
    Anonymous,
    Authorized(Authorized),
}

impl Auth {
    pub fn authenticate_http(
        &self,
        auth_header: Option<&hyper::header::HeaderValue>,
        disable_namespaces: bool,
        namespace_jwt_key: Option<jsonwebtoken::DecodingKey>,
    ) -> Result<Authenticated, AuthError> {
        if self.disabled {
            return Ok(Authenticated::Authorized(Authorized {
                namespace: None,
                permission: Permission::FullAccess,
            }));
        }

        let Some(auth_header) = auth_header else {
            return Err(AuthError::HttpAuthHeaderMissing);
        };

        match parse_http_auth_header(auth_header)? {
            HttpAuthHeader::Basic(actual_value) => {
                let Some(expected_value) = self.http_basic.as_ref() else {
                    return Err(AuthError::BasicNotAllowed);
                };
                // NOTE: this naive comparison may leak information about the `expected_value`
                // using a timing attack
                let actual_value = actual_value.trim_end_matches('=');
                let expected_value = expected_value.trim_end_matches('=');
                if actual_value == expected_value {
                    Ok(Authenticated::Authorized(Authorized {
                        namespace: None,
                        permission: Permission::FullAccess,
                    }))
                } else {
                    Err(AuthError::BasicRejected)
                }
            }
            HttpAuthHeader::Bearer(token) => {
                self.validate_jwt(&token, disable_namespaces, namespace_jwt_key)
            }
        }
    }

    pub fn authenticate_grpc<T>(
        &self,
        req: &tonic::Request<T>,
        disable_namespaces: bool,
        namespace_jwt_key: Option<jsonwebtoken::DecodingKey>,
    ) -> Result<Authenticated, Status> {
        let metadata = req.metadata();

        let auth = metadata
            .get(GRPC_AUTH_HEADER)
            .map(|v| v.to_bytes().expect("Auth should always be ASCII"))
            .map(|v| HeaderValue::from_maybe_shared(v).expect("Should already be valid header"));

        self.authenticate_http(auth.as_ref(), disable_namespaces, namespace_jwt_key)
            .map_err(Into::into)
    }

    pub fn authenticate_jwt(
        &self,
        jwt: Option<&str>,
        disable_namespaces: bool,
        namespace_jwt_key: Option<jsonwebtoken::DecodingKey>,
    ) -> Result<Authenticated, AuthError> {
        if self.disabled {
            return Ok(Authenticated::Authorized(Authorized {
                namespace: None,
                permission: Permission::FullAccess,
            }));
        }

        let Some(jwt) = jwt else {
            return Err(AuthError::JwtMissing);
        };

        self.validate_jwt(jwt, disable_namespaces, namespace_jwt_key)
    }

    fn validate_jwt(
        &self,
        jwt: &str,
        disable_namespaces: bool,
        namespace_jwt_key: Option<jsonwebtoken::DecodingKey>,
    ) -> Result<Authenticated, AuthError> {
        let jwt_key = match namespace_jwt_key.as_ref() {
            Some(jwt_key) => jwt_key,
            None => match self.jwt_key.as_ref() {
                Some(jwt_key) => jwt_key,
                None => return Err(AuthError::JwtNotAllowed),
            },
        };
        validate_jwt(jwt_key, jwt, disable_namespaces)
    }
}

impl Authenticated {
    pub fn from_proxy_grpc_request<T>(
        req: &tonic::Request<T>,
        disable_namespace: bool,
    ) -> Result<Self, Status> {
        let namespace = if disable_namespace {
            None
        } else {
            req.metadata()
                .get_bin(NAMESPACE_METADATA_KEY)
                .map(|c| c.to_bytes())
                .transpose()
                .map_err(|_| Status::invalid_argument("failed to parse namespace header"))?
                .map(NamespaceName::from_bytes)
                .transpose()
                .map_err(|_| Status::invalid_argument("invalid namespace name"))?
        };

        let auth = match req
            .metadata()
            .get(GRPC_PROXY_AUTH_HEADER)
            .map(|v| v.to_str())
            .transpose()
            .map_err(|_| Status::invalid_argument("missing authorization header"))?
        {
            Some("full_access") => Authenticated::Authorized(Authorized {
                namespace,
                permission: Permission::FullAccess,
            }),
            Some("read_only") => Authenticated::Authorized(Authorized {
                namespace,
                permission: Permission::ReadOnly,
            }),
            Some("anonymous") => Authenticated::Anonymous,
            Some(level) => {
                return Err(Status::permission_denied(format!(
                    "invalid authorization level: {}",
                    level
                )))
            }
            None => return Err(Status::invalid_argument("x-proxy-authorization not set")),
        };

        Ok(auth)
    }

    pub fn upgrade_grpc_request<T>(&self, req: &mut tonic::Request<T>) {
        let key = tonic::metadata::AsciiMetadataKey::from_static(GRPC_PROXY_AUTH_HEADER);

        let auth = match self {
            Authenticated::Anonymous => "anonymous",
            Authenticated::Authorized(Authorized {
                permission: Permission::FullAccess,
                ..
            }) => "full_access",
            Authenticated::Authorized(Authorized {
                permission: Permission::ReadOnly,
                ..
            }) => "read_only",
        };

        let value = tonic::metadata::AsciiMetadataValue::try_from(auth).unwrap();

        req.metadata_mut().insert(key, value);
    }

    pub fn is_namespace_authorized(&self, namespace: &NamespaceName) -> bool {
        match self {
            Authenticated::Anonymous => false,
            Authenticated::Authorized(Authorized {
                namespace: Some(ns),
                ..
            }) => ns == namespace,
            // we threat the absence of a specific namespace has a permission to any namespace
            Authenticated::Authorized(Authorized {
                namespace: None, ..
            }) => true,
        }
    }
}

#[derive(Debug)]
enum HttpAuthHeader {
    Basic(String),
    Bearer(String),
}

fn parse_http_auth_header(
    header: &hyper::header::HeaderValue,
) -> Result<HttpAuthHeader, AuthError> {
    let Ok(header) = header.to_str() else {
        return Err(AuthError::HttpAuthHeaderInvalid);
    };

    let Some((scheme, param)) = header.split_once(' ') else {
        return Err(AuthError::HttpAuthHeaderInvalid);
    };

    if scheme.eq_ignore_ascii_case("basic") {
        Ok(HttpAuthHeader::Basic(param.into()))
    } else if scheme.eq_ignore_ascii_case("bearer") {
        Ok(HttpAuthHeader::Bearer(param.into()))
    } else {
        Err(AuthError::HttpAuthHeaderUnsupportedScheme)
    }
}

fn validate_jwt(
    jwt_key: &jsonwebtoken::DecodingKey,
    jwt: &str,
    disable_namespace: bool,
) -> Result<Authenticated, AuthError> {
    use jsonwebtoken::errors::ErrorKind;

    let mut validation = jsonwebtoken::Validation::new(jsonwebtoken::Algorithm::EdDSA);
    validation.required_spec_claims.remove("exp");

    match jsonwebtoken::decode::<serde_json::Value>(jwt, jwt_key, &validation).map(|t| t.claims) {
        Ok(serde_json::Value::Object(claims)) => {
            tracing::trace!("Claims: {claims:#?}");
            let namespace = if disable_namespace {
                None
            } else {
                claims
                    .get("id")
                    .and_then(|ns| NamespaceName::from_string(ns.as_str()?.into()).ok())
            };

            let permission = match claims.get("a").and_then(|s| s.as_str()) {
                Some("ro") => Permission::ReadOnly,
                Some("rw") => Permission::FullAccess,
                Some(_) => return Ok(Authenticated::Anonymous),
                // Backward compatibility - no access claim means full access
                None => Permission::FullAccess,
            };

            Ok(Authenticated::Authorized(Authorized {
                namespace,
                permission,
            }))
        }
        Ok(_) => Err(AuthError::JwtInvalid),
        Err(error) => Err(match error.kind() {
            ErrorKind::InvalidToken
            | ErrorKind::InvalidSignature
            | ErrorKind::InvalidAlgorithm
            | ErrorKind::Base64(_)
            | ErrorKind::Json(_)
            | ErrorKind::Utf8(_) => AuthError::JwtInvalid,
            ErrorKind::ExpiredSignature => AuthError::JwtExpired,
            ErrorKind::ImmatureSignature => AuthError::JwtImmature,
            _ => AuthError::Other,
        }),
    }
}

pub fn parse_http_basic_auth_arg(arg: &str) -> Result<Option<String>> {
    if arg == "always" {
        return Ok(None);
    }

    let Some((scheme, param)) = arg.split_once(':') else {
        bail!("invalid HTTP auth config: {arg}")
    };

    if scheme == "basic" {
        Ok(Some(param.into()))
    } else {
        bail!("unsupported HTTP auth scheme: {scheme:?}")
    }
}

pub fn parse_jwt_key(data: &str) -> Result<jsonwebtoken::DecodingKey> {
    if data.starts_with("-----BEGIN PUBLIC KEY-----") {
        jsonwebtoken::DecodingKey::from_ed_pem(data.as_bytes())
            .context("Could not decode Ed25519 public key from PEM")
    } else if data.starts_with("-----BEGIN PRIVATE KEY-----") {
        bail!("Received a private key, but a public key is expected")
    } else if data.starts_with("-----BEGIN") {
        bail!("Key is in unsupported PEM format")
    } else {
        jsonwebtoken::DecodingKey::from_ed_components(data)
            .context("Could not decode Ed25519 public key from base64")
    }
}

impl AuthError {
    pub fn code(&self) -> &'static str {
        match self {
            Self::HttpAuthHeaderMissing => "AUTH_HTTP_HEADER_MISSING",
            Self::HttpAuthHeaderInvalid => "AUTH_HTTP_HEADER_INVALID",
            Self::HttpAuthHeaderUnsupportedScheme => "AUTH_HTTP_HEADER_UNSUPPORTED_SCHEME",
            Self::BasicNotAllowed => "AUTH_BASIC_NOT_ALLOWED",
            Self::BasicRejected => "AUTH_BASIC_REJECTED",
            Self::JwtMissing => "AUTH_JWT_MISSING",
            Self::JwtNotAllowed => "AUTH_JWT_NOT_ALLOWED",
            Self::JwtInvalid => "AUTH_JWT_INVALID",
            Self::JwtExpired => "AUTH_JWT_EXPIRED",
            Self::JwtImmature => "AUTH_JWT_IMMATURE",
            Self::Other => "AUTH_FAILED",
        }
    }
}

impl From<AuthError> for Status {
    fn from(e: AuthError) -> Self {
        Status::unauthenticated(format!("AuthError: {}", e))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use hyper::header::HeaderValue;

    fn authenticate_http(auth: &Auth, header: &str) -> Result<Authenticated, AuthError> {
        auth.authenticate_http(Some(&HeaderValue::from_str(header).unwrap()), false, None)
    }

    const VALID_JWT_KEY: &str = "zaMv-aFGmB7PXkjM4IrMdF6B5zCYEiEGXW3RgMjNAtc";
    const VALID_JWT: &str = "eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCJ9.\
        eyJleHAiOjc5ODg0ODM4Mjd9.\
        MatB2aLnPFusagqH2RMoVExP37o2GFLmaJbmd52OdLtAehRNeqeJZPrefP1t2GBFidApUTLlaBRL6poKq_s3CQ";
    const VALID_READONLY_JWT: &str = "eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCJ9.\
        eyJleHAiOjc5ODg0ODM4MjcsImEiOiJybyJ9.\
        _2ZZiO2HC8b3CbCHSCufXXBmwpl-dLCv5O9Owvpy7LZ9aiQhXODpgV-iCdTsLQJ5FVanWhfn3FtJSnmWHn25DQ";

    macro_rules! assert_ok {
        ($e:expr) => {
            let res = $e;
            if let Err(err) = res {
                panic!("Expected Ok, got Err({:?})", err)
            }
        };
    }

    macro_rules! assert_err {
        ($e:expr) => {
            let res = $e;
            if let Ok(ok) = res {
                panic!("Expected Err, got Ok({:?})", ok);
            }
        };
    }

    #[test]
    fn test_default() {
        let auth = Auth::default();
        assert_err!(auth.authenticate_http(None, false, None));
        assert_err!(authenticate_http(&auth, "Basic d29qdGVrOnRoZWJlYXI="));
        assert_err!(auth.authenticate_jwt(Some(VALID_JWT), false, None));
    }

    #[test]
    fn test_http_basic() {
        let auth = Auth {
            http_basic: parse_http_basic_auth_arg("basic:d29qdGVrOnRoZWJlYXI=").unwrap(),
            ..Auth::default()
        };
        assert_ok!(authenticate_http(&auth, "Basic d29qdGVrOnRoZWJlYXI="));
        assert_ok!(authenticate_http(&auth, "Basic d29qdGVrOnRoZWJlYXI"));
        assert_ok!(authenticate_http(&auth, "Basic d29qdGVrOnRoZWJlYXI==="));

        assert_ok!(authenticate_http(&auth, "basic d29qdGVrOnRoZWJlYXI="));

        assert_err!(authenticate_http(&auth, "Basic d29qdgvronrozwjlyxi="));
        assert_err!(authenticate_http(&auth, "Basic d29qdGVrOnRoZWZveA=="));

        assert_err!(auth.authenticate_http(None, false, None));
        assert_err!(authenticate_http(&auth, ""));
        assert_err!(authenticate_http(&auth, "foobar"));
        assert_err!(authenticate_http(&auth, "foo bar"));
        assert_err!(authenticate_http(&auth, "basic #$%^"));
    }

    #[test]
    fn test_http_bearer() {
        let auth = Auth {
            jwt_key: Some(parse_jwt_key(VALID_JWT_KEY).unwrap()),
            ..Auth::default()
        };
        assert_ok!(authenticate_http(&auth, &format!("Bearer {VALID_JWT}")));
        assert_ok!(authenticate_http(&auth, &format!("bearer {VALID_JWT}")));

        assert_err!(authenticate_http(&auth, "Bearer foobar"));
        assert_err!(authenticate_http(
            &auth,
            &format!("Bearer {}", &VALID_JWT[..80])
        ));

        assert_eq!(
            authenticate_http(&auth, &format!("Bearer {VALID_READONLY_JWT}")).unwrap(),
            Authenticated::Authorized(Authorized {
                namespace: None,
                permission: Permission::ReadOnly
            })
        );
    }

    #[test]
    fn test_jwt() {
        let auth = Auth {
            jwt_key: Some(parse_jwt_key(VALID_JWT_KEY).unwrap()),
            ..Auth::default()
        };
        assert_ok!(auth.authenticate_jwt(Some(VALID_JWT), false, None));
        assert_err!(auth.authenticate_jwt(Some(&VALID_JWT[..80]), false, None));
    }
}
