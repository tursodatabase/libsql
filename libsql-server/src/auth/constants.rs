pub(crate) static GRPC_AUTH_HEADER: &str = "x-authorization";
pub(crate) static GRPC_PROXY_AUTH_HEADER: &str = "x-proxy-authorization";
