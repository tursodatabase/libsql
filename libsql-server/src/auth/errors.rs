use tonic::Status;

#[derive(thiserror::Error, Debug, PartialEq)]
pub enum AuthError {
    #[error("The `Authorization` HTTP header is required but was not specified")]
    HttpAuthHeaderMissing,
    #[error("The `Authorization` HTTP header has invalid value")]
    HttpAuthHeaderInvalid,
    #[error("The authentication scheme in the `Authorization` HTTP header is not supported")]
    HttpAuthHeaderUnsupportedScheme,
    #[error("The `Basic` HTTP authentication scheme is not allowed")]
    BasicNotAllowed,
    #[error("The `Basic` HTTP authentication credentials were rejected")]
    BasicRejected,
    #[error("Authentication is required but no JWT was specified")]
    JwtMissing,
    #[error("Authentication using a JWT is not allowed")]
    JwtNotAllowed,
    #[error("The JWT is invalid")]
    JwtInvalid,
    #[error("The JWT has expired")]
    JwtExpired,
    #[error("The JWT is immature (not valid yet)")]
    JwtImmature,
    #[error("Auth string does not conform to '<scheme> <token>' form")]
    AuthStringMalformed,
    #[error("Expected authorization header but none given")]
    AuthHeaderNotFound,
    #[error("Expected authorization proxy header but none given")]
    AuthProxyHeaderNotFound,
    #[error("Failed to parse auth proxy header")]
    AuthProxyHeaderInvalid,
    #[error("Non-ASCII auth header")]
    AuthHeaderNonAscii,
    #[error("Authentication failed")]
    Other,
}

impl AuthError {
    pub fn code(&self) -> &'static str {
        match self {
            Self::HttpAuthHeaderMissing => "AUTH_HTTP_HEADER_MISSING",
            Self::HttpAuthHeaderInvalid => "AUTH_HTTP_HEADER_INVALID",
            Self::HttpAuthHeaderUnsupportedScheme => "AUTH_HTTP_HEADER_UNSUPPORTED_SCHEME",
            Self::BasicNotAllowed => "AUTH_BASIC_NOT_ALLOWED",
            Self::BasicRejected => "AUTH_BASIC_REJECTED",
            Self::JwtMissing => "AUTH_JWT_MISSING",
            Self::JwtNotAllowed => "AUTH_JWT_NOT_ALLOWED",
            Self::JwtInvalid => "AUTH_JWT_INVALID",
            Self::JwtExpired => "AUTH_JWT_EXPIRED",
            Self::JwtImmature => "AUTH_JWT_IMMATURE",
            Self::AuthStringMalformed => "AUTH_HEADER_MALFORMED",
            Self::AuthHeaderNotFound => "AUTH_HEADER_NOT_FOUND",
            Self::AuthProxyHeaderNotFound => "AUTH_PROXY_HEADER_NOT_FOUND",
            Self::AuthProxyHeaderInvalid => "AUTH_PROXY_HEADER_INVALID",
            Self::AuthHeaderNonAscii => "AUTH_HEADER_MALFORMED",
            Self::Other => "AUTH_FAILED",
        }
    }
}

impl From<AuthError> for Status {
    fn from(e: AuthError) -> Self {
        Status::unauthenticated(format!("AuthError: {}", e))
    }
}
