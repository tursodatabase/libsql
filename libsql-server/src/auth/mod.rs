use std::sync::Arc;

pub mod authenticated;
pub mod authorized;
pub mod constants;
pub mod errors;
pub mod parsers;
pub mod permission;
pub mod user_auth_strategies;

pub use authenticated::Authenticated;
pub use authorized::Authorized;
pub use errors::AuthError;
pub use parsers::{parse_http_auth_header, parse_http_basic_auth_arg, parse_jwt_key};
pub use permission::Permission;
pub use user_auth_strategies::{
    Disabled, HttpBasic, Jwt, ProxyGrpc, UserAuthContext, UserAuthStrategy,
};

#[derive(Clone)]
pub struct Auth {
    pub strategy: Arc<dyn UserAuthStrategy + Send + Sync>,
}

impl Auth {
    pub fn new(strategy: impl UserAuthStrategy + Send + Sync + 'static) -> Self {
        Self {
            strategy: Arc::new(strategy),
        }
    }

    pub fn authenticate(&self, context: UserAuthContext) -> Result<Authenticated, AuthError> {
        self.strategy.authenticate(context)
    }
}
