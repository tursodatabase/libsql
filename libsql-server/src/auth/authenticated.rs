use std::sync::Arc;

use crate::auth::{constants::GRPC_PROXY_AUTH_HEADER, Authorized};
use crate::namespace::NamespaceName;
use tonic::Status;

use super::authorized::Scope;
use super::Permission;

/// A witness that the user has been authenticated.
#[non_exhaustive]
#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
pub enum Authenticated {
    Anonymous,
    Authorized(Arc<Authorized>),
    Legacy(LegacyAuth),
    FullAccess,
}

#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
pub struct LegacyAuth {
    pub(crate) namespace: Option<NamespaceName>,
    pub(crate) perm: Permission,
}

impl Authenticated {
    pub fn from_proxy_grpc_request<T>(req: &tonic::Request<T>) -> Result<Self, Status> {
        let auth = match req
            .metadata()
            .get(GRPC_PROXY_AUTH_HEADER)
            .map(|v| v.to_str())
            .transpose()
            .map_err(|_| Status::invalid_argument("missing authorization header"))?
        {
            Some(s) => serde_json::from_str::<Authenticated>(s).unwrap(),
            None => return Err(Status::invalid_argument("x-proxy-authorization not set")),
        };

        Ok(auth)
    }

    pub fn upgrade_grpc_request<T>(&self, req: &mut tonic::Request<T>) {
        let key = tonic::metadata::AsciiMetadataKey::from_static(GRPC_PROXY_AUTH_HEADER);

        let auth = serde_json::to_string(self).unwrap();
        let value = tonic::metadata::AsciiMetadataValue::try_from(auth).unwrap();

        req.metadata_mut().insert(key, value);
    }

    pub fn is_namespace_authorized(&self, namespace: &NamespaceName) -> bool {
        match self {
            Authenticated::Anonymous => false,
            Authenticated::Authorized(auth) => {
                auth.has_right(Scope::Namespace(namespace.clone()), Permission::Read)
            }
            Authenticated::FullAccess => true,
            Authenticated::Legacy(auth) => {
                auth.namespace.is_none() || auth.namespace.iter().any(|ns| ns == namespace)
            }
        }
    }

    pub(crate) fn has_right(
        &self,
        namespace: &NamespaceName,
        perm: Permission,
    ) -> crate::Result<()> {
        match self {
            Authenticated::Anonymous => Err(crate::Error::NotAuthorized(
                "anonymous access not allowed".to_string(),
            )),
            Authenticated::Authorized(a) => {
                if !a.has_right(Scope::Namespace(namespace.clone()), perm) {
                    Err(crate::Error::NotAuthorized(format!(
                                "Current session doesn't not have {perm:?} permission to namespace {namespace}")))
                } else {
                    Ok(())
                }
            }
            Authenticated::FullAccess => Ok(()),
            Authenticated::Legacy(auth) => {
                if self.is_namespace_authorized(namespace) && Permission::has_right(auth.perm, perm)
                {
                    Ok(())
                } else {
                    Err(crate::Error::NotAuthorized(format!(
                                "Current session doesn't not have {perm:?} permission to namespace {namespace}")))
                }
            }
        }
    }

    pub(crate) fn ddl_permitted(&self, namespace: &NamespaceName) -> crate::Result<()> {
        match self {
            Authenticated::Authorized(a) if a.ddl_permitted(namespace) => Ok(()),
            Authenticated::FullAccess => Ok(()),
            _ => Err(crate::Error::NotAuthorized(format!(
                "DDL statements not permitted on namespace {namespace}"
            ))),
        }
    }
}
