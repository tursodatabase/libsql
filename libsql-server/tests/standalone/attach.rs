use std::time::Duration;

use insta::assert_debug_snapshot;
use libsql::Database;
use uuid::Uuid;

use crate::common::auth::{encode, key_pair};
use crate::common::http::Client;
use crate::common::net::TurmoilConnector;

use super::make_standalone_server;

#[test]
fn attach_no_auth() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();

    sim.host("primary", make_standalone_server);

    sim.client("test", async {
        let client = Client::new();

        client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                serde_json::json!({}),
            )
            .await
            .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/bar/create",
                serde_json::json!({ "allow_attach": true}),
            )
            .await
            .unwrap();

        let foo_db =
            Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo_db.connect().unwrap();
        foo_conn
            .execute("CREATE TABLE foo_table (x)", ())
            .await
            .unwrap();
        foo_conn
            .execute("insert into foo_table values (42)", ())
            .await
            .unwrap();

        let bar_db =
            Database::open_remote_with_connector("http://bar.primary:8080", "", TurmoilConnector)?;
        let bar_conn = bar_db.connect().unwrap();
        bar_conn
            .execute("CREATE TABLE bar_table (x)", ())
            .await
            .unwrap();
        bar_conn
            .execute("insert into bar_table values (43)", ())
            .await
            .unwrap();

        // fails: foo doesn't allow attach
        assert_debug_snapshot!(bar_conn.execute("ATTACH foo as foo", ()).await.unwrap_err());

        let txn = foo_conn.transaction().await.unwrap();
        txn.execute("ATTACH DATABASE bar as bar", ()).await.unwrap();
        let mut rows = txn.query("SELECT * FROM bar.bar_table", ()).await.unwrap();
        // succeeds!
        assert_debug_snapshot!(rows.next().await);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn attach_auth() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();

    sim.host("primary", make_standalone_server);

    sim.client("test", async {
        let client = Client::new();

        let (enc, jwt_key) = key_pair();

        assert!(client
            .post(
                "http://primary:9090/v1/namespaces/foo/create",
                serde_json::json!({ "jwt_key": jwt_key })
            )
            .await
            .unwrap()
            .status()
            .is_success());
        assert!(client
            .post(
                "http://primary:9090/v1/namespaces/bar/create",
                serde_json::json!({ "allow_attach": true, "jwt_key": jwt_key })
            )
            .await
            .unwrap()
            .status()
            .is_success());

        let claims = serde_json::json!({
            "p": {
                "rw": {
                    "ns": ["bar", "foo"]
                }
            }
        });
        let token = encode(&claims, &enc);

        let foo_db = Database::open_remote_with_connector(
            "http://foo.primary:8080",
            &token,
            TurmoilConnector,
        )?;
        let foo_conn = foo_db.connect().unwrap();
        foo_conn
            .execute("CREATE TABLE foo_table (x)", ())
            .await
            .unwrap();
        foo_conn
            .execute("insert into foo_table values (42)", ())
            .await
            .unwrap();

        let bar_db = Database::open_remote_with_connector(
            "http://bar.primary:8080",
            &token,
            TurmoilConnector,
        )?;
        let bar_conn = bar_db.connect().unwrap();
        bar_conn
            .execute("CREATE TABLE bar_table (x)", ())
            .await
            .unwrap();
        bar_conn
            .execute("insert into bar_table values (43)", ())
            .await
            .unwrap();

        // fails: no perm
        assert_debug_snapshot!(bar_conn.execute("ATTACH foo as foo", ()).await.unwrap_err());

        let txn = foo_conn.transaction().await.unwrap();
        // fails: no perm
        assert_debug_snapshot!(txn
            .execute("ATTACH DATABASE bar as bar", ())
            .await
            .unwrap_err());

        let claims = serde_json::json!({
            "p": {
                "roa": {
                    "ns": ["bar", "foo"]
                }
            }
        });
        let token = encode(&claims, &enc);

        let foo_db = Database::open_remote_with_connector(
            "http://foo.primary:8080",
            &token,
            TurmoilConnector,
        )?;
        let foo_conn = foo_db.connect().unwrap();
        let bar_db = Database::open_remote_with_connector(
            "http://bar.primary:8080",
            &token,
            TurmoilConnector,
        )?;
        let bar_conn = bar_db.connect().unwrap();

        // fails: namesapce doesn't allow attach
        assert_debug_snapshot!(bar_conn.execute("ATTACH foo as foo", ()).await.unwrap_err());

        let txn = foo_conn.transaction().await.unwrap();
        txn.execute("ATTACH DATABASE bar as bar", ()).await.unwrap();
        let mut rows = txn.query("SELECT * FROM bar.bar_table", ()).await.unwrap();
        // succeeds!
        assert_debug_snapshot!(rows.next().await);

        // mixed claims
        let claims = serde_json::json!({
            "p": {
                "rw": {
                    "ns": ["foo"]
                },
                "roa": {
                    "ns": ["bar"]
                }
            }
        });
        let token = encode(&claims, &enc);

        let foo_db = Database::open_remote_with_connector(
            "http://foo.primary:8080",
            &token,
            TurmoilConnector,
        )?;
        let foo_conn = foo_db.connect().unwrap();
        let txn = foo_conn.transaction().await.unwrap();
        txn.execute("ATTACH DATABASE bar as attached", ())
            .await
            .unwrap();
        let mut rows = txn
            .query("SELECT * FROM attached.bar_table", ())
            .await
            .unwrap();
        // succeeds!
        assert_debug_snapshot!(rows.next().await);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn attach_auth_with_uuids() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();

    sim.host("primary", make_standalone_server);

    sim.client("test", async {
        let client = Client::new();

        let (enc, jwt_key) = key_pair();

        let main_db_id = Uuid::new_v4();
        let attach_db_id = Uuid::new_v4();

        assert!(client
            .post(
                format!("http://primary:9090/v1/namespaces/{}/create", main_db_id).as_str(),
                serde_json::json!({ "jwt_key": jwt_key })
            )
            .await
            .unwrap()
            .status()
            .is_success());
        assert!(client
            .post(
                format!("http://primary:9090/v1/namespaces/{}/create", attach_db_id).as_str(),
                serde_json::json!({ "allow_attach": true, "jwt_key": jwt_key })
            )
            .await
            .unwrap()
            .status()
            .is_success());

        let claims = serde_json::json!({
            "p": {
                "rw": {
                    "ns": [main_db_id, attach_db_id]
                },
                "roa": {
                    "ns": [attach_db_id]
                }
            }
        });
        let token = encode(&claims, &enc);

        let attach_conn = Database::open_remote_with_connector(
            format!("http://{}.primary:8080", attach_db_id).as_str(),
            &token,
            TurmoilConnector,
        )?
        .connect()
        .unwrap();
        attach_conn
            .execute("CREATE TABLE bar_table (x)", ())
            .await
            .unwrap();
        attach_conn
            .execute("insert into bar_table values (43)", ())
            .await
            .unwrap();

        let main_conn = Database::open_remote_with_connector(
            format!("http://{}.primary:8080", main_db_id).as_str(),
            &token,
            TurmoilConnector,
        )?
        .connect()
        .unwrap();

        // fails: namespace is uuid, hence needs to be wrapped in quotes
        assert_debug_snapshot!(main_conn
            .execute(
                "ATTACH DATABASE ae308915-caca-480f-a6b4-9f9f9dc84b11 as bar",
                ()
            )
            .await
            .unwrap_err());

        let txn = main_conn.transaction().await.unwrap();
        txn.execute(
            format!("ATTACH DATABASE \"{}\" as bar", attach_db_id).as_str(),
            (),
        )
        .await
        .unwrap();
        let mut rows = txn.query("SELECT * FROM bar.bar_table", ()).await.unwrap();
        assert_debug_snapshot!(rows.next().await);
        Ok(())
    });

    sim.run().unwrap();
}
