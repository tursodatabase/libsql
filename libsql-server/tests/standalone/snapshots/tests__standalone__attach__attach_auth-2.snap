---
source: libsql-server/tests/standalone/attach.rs
expression: "txn.execute(\"ATTACH DATABASE bar as bar\", ()).await.unwrap_err()"
---
Hrana(
    Api(
        "{\"error\":\"Internal Error: `Not authorized to execute query: Current session doesn't not have AttachRead permission to namespace bar`\"}",
    ),
)
